packages/get-starknet/src/wallet.ts

```diff
@@ -220,13 +236,108 @@ export class MetaMaskSnapWallet implements StarknetWindowObject {
     return true;
   }
 
-  // eslint-disable-next-line @typescript-eslint/no-unused-vars
-  on<Event extends keyof WalletEventHandlers>(_event: Event, _handleEvent: WalletEventHandlers[Event]): void {
-    // No operation for now
+  /**
+   * Registers an event handler for `accountsChanged` or `networkChanged`.
+   *
+   * @param event - The event name ('accountsChanged' or 'networkChanged')
+   * @param handleEvent - The event handler function
+   */
+  on<Event extends keyof WalletEventHandlers>(event: Event, handleEvent: WalletEventHandlers[Event]): void {
+    if (!this.#pollingController) {
+      this.#startPolling();
+    }
+
+    if (event === 'accountsChanged') {
+      this.#accountChangeHandlers.push(handleEvent as AccountChangeEventHandler);
+    } else if (event === 'networkChanged') {
+      this.#networkChangeHandlers.push(handleEvent as NetworkChangeEventHandler);
+    } else {
+      throw new Error(`Unsupported event: ${String(event)}`);
+    }
+  }
+
+  /**
+   * Removes an event handler for `accountsChanged` or `networkChanged`.
+   *
+   * @param event - The event name ('accountsChanged' or 'networkChanged')
+   * @param _handleEvent
+   */
+  off<Event extends keyof WalletEventHandlers>(event: Event, _handleEvent?: WalletEventHandlers[Event]): void {
+    if (event === 'accountsChanged') {
+      this.#accountChangeHandlers = [];
+    } else if (event === 'networkChanged') {
+      this.#networkChangeHandlers = [];
+    } else {
+      throw new Error(`Unsupported event: ${String(event)}`);
+    }
+
+    if (this.#accountChangeHandlers.length === 0 && this.#networkChangeHandlers.length === 0) {
+      this.#stopPolling();
+    }
+  }
+
+  /**
+   * Polling function for detecting changes with a maximum delay.
+   * Ensures that each iteration respects a maximum delay before proceeding.
+   */
+  #pollingFunction = async (): Promise<void> => {
+    if (!this.#pollingController) {
+      return; // Abort if the polling controller is not initialized
+    }
+    const { signal } = this.#pollingController;
+
+    while (!signal.aborted) {
+      const startTime = Date.now(); // Track the start time of the iteration
+      const previousNetwork = this.#network;
```

Comment:
nit:
we have chain id in the class
```suggestion
      const previousNetwork = this.#chainid;
```